dbms/src/IO/Compression/CompressionCodecFor.cpp

```diff
@@ -0,0 +1,227 @@
+// Copyright 2024 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <Common/BitpackingPrimitives.h>
+#include <Common/Exception.h>
+#include <IO/Compression/CompressionCodecFor.h>
+#include <IO/Compression/CompressionInfo.h>
+#include <common/unaligned.h>
+
+#if defined(__AVX2__)
+#include <immintrin.h>
+#endif
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+extern const int CANNOT_COMPRESS;
+extern const int CANNOT_DECOMPRESS;
+} // namespace ErrorCodes
+
+CompressionCodecFor::CompressionCodecFor(UInt8 bytes_size_)
+    : bytes_size(bytes_size_)
+{}
+
+UInt8 CompressionCodecFor::getMethodByte() const
+{
+    return static_cast<UInt8>(CompressionMethodByte::For);
+}
+
+UInt32 CompressionCodecFor::getMaxCompressedDataSize(UInt32 uncompressed_size) const
+{
+    // 1 byte for bytes_size, x bytes for frame of reference, 1 byte for width.
+    size_t count = uncompressed_size / bytes_size;
+    return 1 + bytes_size + sizeof(UInt8) + BitpackingPrimitives::getRequiredSize(count, bytes_size * 8);
+}
+
+template <typename T>
+UInt32 CompressionCodecFor::compressData(const char * source, UInt32 source_size, char * dest)
+{
+    static_assert(std::is_integral<T>::value, "Integral required.");
+    const auto count = source_size / sizeof(T);
+    std::vector<T> values;
+    values.reserve(count);
+    const char * const source_end = source + source_size;
+    while (source < source_end)
+    {
+        values.push_back(unalignedLoad<T>(source));
+        source += sizeof(T);
+    }
+    T frame_of_reference = *std::min_element(values.cbegin(), values.cend());
+    // store frame of reference
+    unalignedStore<T>(dest, frame_of_reference);
+    dest += sizeof(T);
+    if (frame_of_reference != 0)
```

Comment:
T can only be an unsigned type, `frame_of_reference` can not be negative.

Reply:
Now we support signed types, but here is just an optimization, when `frame_of_reference` = 0, no need to do minus.

And for signed types, there is no need to consider overflow, because

```cpp
int count = std::numeric_limits<int>::min();
count -= std::numeric_limits<int>::max();
count += std::numeric_limits<int>::max();
std::cout << count << "\n"; // equal to std::numeric_limits<int>::min()
```